packages/nitro-protocol/contracts/MultiAssetHolder.sol

```diff
@@ -2,104 +2,220 @@
+
+        amountDeposited = expectedHeld.add(amount).sub(holdings[asset][destination]); // strictly positive
+        // require successful deposit before updating holdings (protect against reentrancy)
+        if (asset == address(0)) {
+            require(msg.value == amount, 'Incorrect msg.value for deposit');
+        } else {
+            // require successful deposit before updating holdings (protect against reentrancy)
+            require(
+                IERC20(asset).transferFrom(msg.sender, address(this), amountDeposited),
+                'Could not deposit ERC20s'
+            );
+        }
+
+        holdings[asset][destination] = holdings[asset][destination].add(amountDeposited);
+        emit Deposited(destination, asset, amountDeposited, holdings[asset][destination]);
+
+        if (asset == address(0)) {
+            // refund whatever wasn't deposited.
+            uint256 refund = amount.sub(amountDeposited);
+            (bool success, ) = msg.sender.call{value: refund}(''); //solhint-disable-line avoid-low-level-calls
+            require(success, 'Could not refund excess funds');
+        }
+    }
+
     /**
      * @notice Transfers as many funds escrowed against `channelId` as can be afforded for a specific destination. Assumes no repeated entries.
      * @dev Transfers as many funds escrowed against `channelId` as can be afforded for a specific destination. Assumes no repeated entries.
      * @param fromChannelId Unique identifier for state channel to transfer funds *from*.
-     * @param allocationBytes The abi.encode of AssetOutcome.Allocation
+     * @param outcomeBytes The encoded Outcome of this state channel
      * @param indices Array with each entry denoting the index of a destination to transfer funds to. An empty array indicates "all".
      */
     function transfer(
+        uint256 assetIndex, // TODO consider a uint48?
         bytes32 fromChannelId,
-        bytes calldata allocationBytes,
+        bytes memory outcomeBytes,
+        bytes32 stateHash,
+        address challengerAddress,
         uint256[] memory indices
-    ) external override {
+    ) public override returns (bytes memory) {
         // checks
         _requireIncreasingIndices(indices);
-        _requireCorrectAllocationHash(fromChannelId, allocationBytes);
+        _requireChannelFinalized(fromChannelId);
+        _requireMatchingFingerprint(
+            stateHash,
+            challengerAddress,
+            keccak256(outcomeBytes),
+            fromChannelId
+        );
+
+        Outcome.OutcomeItem[] memory outcome = abi.decode(outcomeBytes, (Outcome.OutcomeItem[]));
+        Outcome.AssetOutcome memory assetOutcome = abi.decode(
+            outcome[assetIndex].assetOutcomeBytes,
+            (Outcome.AssetOutcome)
+        );
+        require(
+            assetOutcome.assetOutcomeType == Outcome.AssetOutcomeType.Allocation,
+            '!allocation'
```

Comment:
This is probably copy/pasted, but might be nice to have a friendlier message.
```suggestion
            'Outcome type must be an allocation'
```

Reply:
I think this would trip our linting rules, since we are aiming to have reasons strings shorter than 32 characters (for gas reasons). There is probably something shorter that is still friendly, though. In future we could move to having error codes instead of messages.